e2e/cluster.go

```diff
@@ -132,3 +144,181 @@ func getClusterSubnetID(ctx context.Context, cloud *azureClient, location, resou
+	clusters := []*armcontainerservice.ManagedCluster{}
+	pager := cloud.resourceClient.NewListByResourceGroupPager(resourceGroupName, nil)
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to advance page: %q", err)
+		}
+		for _, resource := range page.Value {
+			if strings.EqualFold(*resource.Type, managedClusterResourceType) {
+				cluster, err := cloud.aksClient.Get(ctx, resourceGroupName, *resource.Name, nil)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get aks cluster: %q", err)
+				}
+				if cluster.Properties == nil {
+					return nil, fmt.Errorf("aks cluster properties were nil")
+				}
+
+				t.Logf("found agentbaker e2e cluster: %q", *cluster.Name)
+				clusters = append(clusters, &cluster.ManagedCluster)
+			}
+		}
+	}
+
+	return clusters, nil
+}
+
+func getViableClusters(scenario *scenario.Scenario, clusters []*armcontainerservice.ManagedCluster) []*armcontainerservice.ManagedCluster {
+	viableClusters := []*armcontainerservice.ManagedCluster{}
+	for _, cluster := range clusters {
+		if scenario.ScenarioConfig.ClusterSelector(cluster) {
+			viableClusters = append(viableClusters, cluster)
+		}
+	}
+	return viableClusters
+}
+
+func mustChooseCluster(
+	ctx context.Context,
+	t *testing.T,
+	r *mrand.Rand,
+	cloud *azureClient,
+	suiteConfig *suiteConfig,
+	scenario *scenario.Scenario,
+	clusters *[]*armcontainerservice.ManagedCluster,
+	paramCache paramCache) (*kubeclient, *armcontainerservice.ManagedCluster, map[string]string, string) {
+	var (
+		chosenKubeClient    *kubeclient
+		chosenCluster       *armcontainerservice.ManagedCluster
+		chosenClusterParams map[string]string
+		chosenSubnetID      string
+	)
+
+	viableClusters := getViableClusters(scenario, *clusters)
+	if len(viableClusters) == 0 {
+		t.Logf("unable to find viable test cluster for scenario %q, attempting to create a new one...", scenario.Name)
+		clusterBaseModel := getBaseClusterModel(
+			fmt.Sprintf(testClusterNameTemplate, randomLowercaseString(r, 5)),
+			suiteConfig.location,
+		)
+		scenario.ScenarioConfig.ClusterMutator(&clusterBaseModel)
+
+		cluster, err := ensureCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, &clusterBaseModel, true)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		kube, subnetID, clusterParams, err := prepareClusterForTests(ctx, t, cloud, suiteConfig, cluster, paramCache)
+		if err != nil {
+			t.Fatalf("unable to prepare new cluster for test: %s", err)
+		}
+
+		*clusters = append(*clusters, cluster)
+
+		chosenCluster = cluster
+		chosenKubeClient = kube
+		chosenSubnetID = subnetID
+		chosenClusterParams = clusterParams
+	} else {
+		for _, cluster := range viableClusters {
+			if _, err := ensureCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, cluster, false); err != nil {
```

Comment:
this fork feels a bit like a code smell (calling ensureCluster in both code paths with isNew true/false but it already exists in one case)

can you share the thinking here?

Reply:
the original thinking was that ensureCluster could be used both to create a new cluster from scratch, and make sure that an existing cluster is in a usable state (MC resource group is there, provisioning state isn't "Failed", etc.) - I guess the smell comes from the fact that this function is probably doing too much without being more clear.

that boolean argument is really just there to short-circuit between "cluster already exists" and "cluster doesn't exist" scenarios, which I guess goes further to suggest that we should split that function up

Reply:
I think the confusing part is maybe the fact that we really have 3 states -- good, dirty, and non-existent.

I'm okay to merge/iterate but would like to break it up a bit more maybe

Reply:
did some splitting